ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/PeerDiscoveryAgent.java

```diff
@@ -178,52 +176,54 @@ protected PeerDiscoveryAgent(
-      final Bytes nodeId,
-      final String advertisedAddress,
-      final Integer tcpPort,
-      final Integer udpPort) {
+  public NodeRecord updateNodeRecord() {
     final KeyValueStorage keyValueStorage =
         storageProvider.getStorageBySegmentIdentifier(KeyValueSegmentIdentifier.BLOCKCHAIN);
     final NodeRecordFactory nodeRecordFactory = NodeRecordFactory.DEFAULT;
     final Optional<NodeRecord> existingNodeRecord =
         keyValueStorage
             .get(Bytes.of(SEQ_NO_STORE_KEY.getBytes(UTF_8)).toArray())
             .map(Bytes::of)
-            .flatMap(b -> Optional.of(nodeRecordFactory.fromBytes(b)));
+            .map(nodeRecordFactory::fromBytes);
 
     final Bytes addressBytes = Bytes.of(InetAddresses.forString(advertisedAddress).getAddress());
-    if (existingNodeRecord.isEmpty()
-        || !existingNodeRecord.get().get(EnrField.PKEY_SECP256K1).equals(nodeId)
-        || !addressBytes.equals(existingNodeRecord.get().get(EnrField.IP_V4))
-        || !tcpPort.equals(existingNodeRecord.get().get(EnrField.TCP))
-        || !udpPort.equals(existingNodeRecord.get().get(EnrField.UDP))) {
-      final UInt64 sequenceNumber =
-          existingNodeRecord.map(NodeRecord::getSeq).orElse(UInt64.ZERO).add(1);
-      final NodeRecord nodeRecord =
-          nodeRecordFactory.createFromValues(
-              sequenceNumber,
-              new EnrField(EnrField.ID, IdentitySchema.V4),
-              new EnrField(EnrField.PKEY_SECP256K1, Functions.compressPublicKey(nodeId)),
-              new EnrField(EnrField.IP_V4, addressBytes),
-              new EnrField(EnrField.TCP, tcpPort),
-              new EnrField(EnrField.UDP, udpPort),
-              new EnrField("eth", Collections.singletonList(forkIdSupplier.get())));
-      nodeRecord.setSignature(
-          nodeKey
-              .sign(Hash.keccak256(nodeRecord.serializeNoSignature()))
-              .encodedBytes()
-              .slice(0, 64));
-
-      final KeyValueStorageTransaction keyValueStorageTransaction =
-          keyValueStorage.startTransaction();
-      keyValueStorageTransaction.put(
-          Bytes.wrap(SEQ_NO_STORE_KEY.getBytes(UTF_8)).toArray(), nodeRecord.serialize().toArray());
-      keyValueStorageTransaction.commit();
-      return Optional.of(nodeRecord);
-    }
-
-    return existingNodeRecord;
+    final Integer discoveryPort = localNode.get().getEnodeURL().getDiscoveryPortOrZero();
+    final Integer listeningPort = localNode.get().getEnodeURL().getListeningPortOrZero();
+    return existingNodeRecord
+        .filter(
+            nodeRecord ->
+                id.equals(nodeRecord.get(EnrField.PKEY_SECP256K1))
+                    && addressBytes.equals(nodeRecord.get(EnrField.IP_V4))
+                    && discoveryPort.equals(nodeRecord.get(EnrField.UDP))
+                    && listeningPort.equals(nodeRecord.get(EnrField.TCP))
+                    && forkIdSupplier.get().equals(nodeRecord.get("eth")))
```

Comment:
non-blocking feedback would be nice to have an enum rather than "eth" magic string.  I see that it is an existing practice though